DSA/tree.cpp

```diff
@@ -0,0 +1,279 @@
+#include <bits/stdc++.h>
+using namespace std;
+
+// Creating Node Class with data, left and right
+
+class Node
+{
+public:
+    int data;
+    Node *left;
+    Node *right;
+
+    Node(int data)
+    {
+        this->data = data;
+        left = NULL;
+        right = NULL;
+    }
+};
+
+// for printing
+
+void print(string message)
+{
+    cout << endl
+         << message;
+}
+
+// for adding element
+
+Node *addElement(Node *&root, int data)
+{
+
+    if (root == NULL)
+    {
+        Node *newNode = new Node(data);
+        root = newNode;
+        return root;
+    }
+
+    if (data > root->data)
+    {
+
+        root->right = addElement(root->right, data);
+    }
+    else
+    {
+        root->left = addElement(root->left, data);
+    }
+
+    return root;
+}
+
+// for adding element the function should not return something
+
+void addVoidEle(Node *&root)
+{
+    // Taking input
+
+    int data;
+    print("Enter the Data: ");
+    cin >> data;
+
+    root = addElement(root, data);
+}
+
+// for deleting element
+
+// Getting mininmum element node from tree
+
+Node *getMin(Node *root)
+{
+    Node *temp = root;
+
+    while (temp->left != NULL)
+    {
+        temp = temp->left;
+    }
+
+    return temp;
+}
+
+// main function to delete node
+
+Node *deleteNode(Node *&root, int key)
+{
+    if (!root)
+    {
+        print("Key does not exist.");
+        return NULL;
+    }
+
+    if (root->data == key)
+    {
+        if (!root->left && !root->right)
+        {
+            return NULL;
+        }
+        else if (root->left && !root->right)
+        {
+            Node *temp = root->left;
+            delete root;
+            return temp;
+        }
+        else if (!root->left && root->right)
+        {
+            Node *temp = root->right;
+            delete root;
+            return temp;
+        }
+        else
+        {
+            Node *min = getMin(root->right);
+            root->data = min->data;
+            root->right = deleteNode(root, min->data);
+
+            return root;
+        }
+    }
+    else
+    {
+        if (root->data > key)
+        {
+            root->left = deleteNode(root->left, key);
+        }
+        else
+        {
+            root->right = deleteNode(root->right, key);
+        }
+    }
+
+    return root;
+}
+
+// a simple funtio to delete which is void
+
+void deleteVoidNode(Node *&root)
+{
+    // Taking input
+
+    int data;
+    print("Enter the data to delete: ");
+    cin >> data;
+
+    root = deleteNode(root, data);
+}
+
+// funtion to search a element in bst
+
+void isExist(Node *root)
+{
+
+    int data;
+    print("Enter the data to search : ");
+    cin >> data;
+
+    Node *temp = root;
```

Comment:
temp named variable naming you should have to avoid . please change it.

Reply:
done